### DISCUSSION TITLE
`nx relese` Better support for integrated monorepos

### DISCUSSION THREAD

**Discussion by groinder** [nrwl/nx#26802] [Feature Requests]
# `nx relese` Better support for integrated monorepos

Love the `nx` lib! It's great that the `nx release` feature is finally here. I'd like to propose an improvement for the integrated monorepos.

## Problem

I have a fairly substantial project at a company I work with. I've set up this project in accordance to the [Integrated Repo](https://nx.dev/concepts/integrated-vs-package-based#integrated-repos) concept and now I'm trying to add the `nx release` to it.

The problem is, that I have a single app and over a dozen of libs scoped to that app, those libs are not publishable - the only reason for their existence is to be used as a part of the app. I also have some libs that started as independent, and ideally I'd like to keep them that way. The `tsconfig.base.json` has all of the paths defined, I can import the libs and everything works great during development, however, it's not so great when it comes to the `nx release`.

I've prepared a testing project to share with you, this is a simplified scenario of our company project (it doesn't include the independent libs).

The config looks as follows:
```
  "release": {
    "projectsRelationship": "independent",
    "projects": ["apps/*"],
    "changelog": {
      "projectChangelogs": true
    },
    "version": {
      "conventionalCommits": true,
      "generatorOptions": {
        "updateDependents": "auto"
      }
    }
  }
```

The dependencies are like this:
![graph](https://github.com/nrwl/nx/assets/16709010/8d58c28b-1d54-4b52-95d2-236166217891)

With such config, If I make a change in a lib that is not versioned - it's just imported in the app, then the `nx release` doesn't pick up those changes, and doesn't trigger a release of the app.

The project can be found here https://github.com/groinder/nx-release-testing, the `main` branch has some extra commits with changes in the `libs`, you can run `npx nx release --dry-run` to see that those are not included in the changelog. There is also [`deps-specified`](https://github.com/groinder/nx-release-testing/tree/deps-specified) branch which shows the second workaround described below.

##  Workarounds

I've found 2 ways to make it work, in both it's required to version the app-scoped libs.

### Common requisites
- Include the libs in `release.projects`.
- Add package.json in all of the libs (it just needs to include the name and the version).

#### Requisites
- Setting the `"projectsRelationship": "fixed",` instead of `independent` - this will bump the release of the app whenever any lib changes.
#### Drawbacks
- All versions will be synced, in some scenarios it may be useful to keep them independent

#### Requisites
- Keeping the `"projectsRelationship": "independent",`.
- Include each lib in the package.json dependencies of **the app**
#### Drawbacks
- With a lot of libs, adding all of them to the dependencies explicitly is tedious (probably can be automated)
- The release process will create a tag for each lib, most of those tags are not needed as what we essentially want is the tag for the app

## Expected solution
Ideally, I'd like it to "just work" out of the box for integrated monorepos regardless of the `projectsRelationship` setting. 
- The `nx release` should detect changes in libs that the app uses, based on the nx graph, whenever a change occurs in such lib it should be included in the changelog for the app.
- Versioning of such libs should not be required - they should be treated as an integral part of the app.
    - no need for adding package.json to each lib
    - no need to include those libs in `release.projects`

## Documentation needed
I've found it hard to find and grasp how exactly to configure the `nx release` for the integrated monorepo setup, regardless if this improvement is made I think it would be useful to include a recipe for this setup in the `nx release` section of the docs.

**Comment by jame-earnin** [comment]
+1 It would be great if Nx also supported the "Application" "affected" release (version bump, changelog, no publish) with integrated base monorepo also. We have a micro-frontend use-case (Module Federation with Nx) where we can release each app independently. We want the ability to bump the version, generate a changelog from the affected graph (and release note), and publish the individual app.
